### PR TITLE
[NVPTX] Only run LowerUnreachable when necessary

### PR DESCRIPTION
**Pull request by justinfargnoli**
Before CUDA 12.3 `ptxas` did not recognize that the trap instruction terminates a basic block. Instead, it would assume that control flow continued to the next instruction. The next instruction could be in the block that's lexically below it. This would lead to phantom CFG edges being created within ptxas. 

[NVPTX: Lower unreachable to exit to allow ptxas to accurately reconstruct the CFG.](https://github.com/llvm/llvm-project/commit/1ee4d880e8760256c606fe55b7af85a4f70d006d) added the LowerUnreachable pass to NVPTX to work around this. Several other WAR patches followed. 

This bug in `ptxas` was fixed in CUDA 12.3 and is thus impossible to encounter when targeting PTX ISA v8.3+

This commit reverts the WARs for the `ptxas` bug when targeting PTX ISA v8.3+

CC @maleadt

### CODE REVIEW

**Comment by Artem-B** [comment]
> this pass is only necessary when targeting Pascal or earlier via ptxas from CUDA 11.4 or earlier. This PR updates NVPTXTargetMachine.cpp to reflect that.

I'm confused. The comment you quoted in the description only says 

> I do still encounter
miscompilations with Julia's unreachable-heavy code when targeting these
older GPUs using an older ptxas version (specifically, from CUDA 11.4 or
below). 

The way I read it, it says that 11.4 and older ptxas with older GPUs may have more bugs that are not fixed by this patch, but I do not think it implies that the newer versions of ptxas or the newer GPUs do not need this pass.

What am I missing?

**Comment by maleadt** [comment]
> The way I read it, it says that 11.4 and older ptxas with older GPUs may have more bugs that are not fixed by this patch

Yeah, that was what I meant.

Although we only know about issues on Pascal and earlier because `bar.sync` is not allowed to be executed divergently there, making sure that `ptxas` has an accurate view of the CFG as intended by LLVM seems important on later hardware generations too. Citing from the relevant NVIDIA bug report (4078847):

> If LLVM creates unstructured control flows then it makes many downstream compiler optimizations less effective.

We could avoid lowering `unreachable` to `exit; trap;` starting from the `ptxas` version that's known to model control flow accurately for just `trap` instructions, which it didn't when I proposed this change, but was told to me would get fixed in a future version of the compiler. I'm not sure if and how we can guard on that though, as the code generated by NVPTX may be handed to whatever version of back-end compiler out there (e.g. when compiling to PTX and using the driver APIs as opposed to having `clang` drive the `ptxas` invocation).

**Comment by Artem-B** [comment]
If we know which ptxas version no longer needs `exit`, then we may use PTX version that's only supported by that version of ptxas (or the driver), or newer. This guarantees that it will never be compileable by the older ptxas with the bug.

```
  bool hasPTXASUnreachableBug() const {
    return SmVersion < 70 && PTXVersion <= 74;
  }
```

With that in mind, this condition should be changed appropriately. Assuming that it's the latest public CUDA release that has the bug fixed, we should only mark PTX ISA 8.5 or newer as safe.  https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#release-notes

**Comment by duk-37** [comment]
> We could avoid lowering unreachable to exit; trap; starting from the ptxas version that's known to model control flow accurately for just trap instructions, which it didn't when I proposed this change, but was told to me would get fixed in a future version of the compiler. I'm not sure if and how we can guard on that though, as the code generated by NVPTX may be handed to whatever version of back-end compiler out there (e.g. when compiling to PTX and using the driver APIs as opposed to having clang drive the ptxas invocation).

That would be a change in the instruction selection backend though, not disabling this pass altogether.

Alternatively, one could just do what WebAssembly and PS4+PS5 do and force these TargetOptions to be enabled. That would get rid of this pass completely. The question of "should TargetOptions be load-bearing" is a good one, but I feel like that bridge has been crossed already and there are many, many other places in which these things are inaccessible or user-provided command line options are just outright ignored.

**Comment by justinfargnoli** [comment]
> I'm confused. The comment you quoted in the description only says

Yes, this is my bad. I should have quoted places where @maleadt said things along the lines of:

> we only know about issues on Pascal and earlier because bar.sync is not allowed to be executed divergently there

---

> Citing from the relevant NVIDIA bug report (4078847):

Thank you for the bug number! I'm following up internally to confirm what release the fix was included in. 

---

> That would be a change in the instruction selection backend though, not disabling this pass altogether

I thought this pass added the `exit` instruction, not instruction selection. Assuming I'm missing something, where should we make this change within the instruction selection? 

> disabling this pass altogether ... mean(s) there's still a need for trap instructions to avoid bogus CFG edges.

This pass doesn't touch the trap instruction, so it should be okay to disable it, right?

**Comment by duk-37** [comment]
> > That would be a change in the instruction selection backend though, not disabling this pass altogether
> 
> I thought this pass added the `exit` instruction, not instruction selection. Assuming I'm missing something, where should we make this change within the instruction selection?
> 

Sorry, I probably added some confusion with the way I worded things. What I was saying is that, if the reason this pass exists is to avoid bogus CFG edges (which makes sense, especially if there are downstream optimizations happening) then it either:
1. should stay as-is, or;
2. `TrapUnreachable` + `NoTrapAfterNoreturrn` could be set to true/false respectively in order to fix the issue in pretty much the same way, just without the extra pass.

The introduction of the pass happened to be due to a miscompile because of bogus CFG edges, but even if that's been fixed it's still probably a useful hint to any compilers taking this output as input like @maleadt said.

**Comment by Artem-B** [comment]
To me it looks like there are two things we need to decide here:

* Can ptxas guarantee that it can handle unstructured CFG in principle? My recollection from the time when we initially struggled wit this issue is that it's not always possible. The nature of "unreachable" code in generated PTX is that it can result in unpredictable phantom CFG edges (as far as ptxas is concerned). Explicitly marking unreachable code with something that allows us to propagate that information to ptxas looks like something that we need to do. `trap`, AFAICT, is the best way to do so that we have right now.

The downside of this approach is that it is an extra instruction and, I believe, there were some reported use cases where it resulted in performance regressions in user code.

Ideally, it would be nice if ptxas would provide a NOP-like equivalent of `unreachable`, which would preserve correctness of its view of CFG without generating any extra instructions.

* generation of `trap; exit;`  is a workaround for the bug in the older ptxas which did not consider `trap` by itself as a CFG leaf. Whether we emit it or not would depend on whether we want/need to emit the unreachable hint instruction to PTX, and, if we do, or on the PTX (or ptxas) version.

**Comment by justinfargnoli** [comment]
> `trap`, AFAICT, is the best way to do so that we have right now.

I followed-up internally and confirmed that `trap` is the best option that we have right now. 

>  if we do, or on the PTX (or ptxas) version.

I've confirmed that this was fixed in CUDA 12.3 which corresponds with PTX version 8.3.

---
> whether we want/need to emit the unreachable hint instruction to PTX

Assuming that we do want to emit the unreachable hint (I don't think anyone has advocated against this), then this PR should be ready for another round of review.

File: llvm/lib/Target/NVPTX/NVPTXTargetMachine.cpp [hunk 1]
```diff
@@ -368,9 +368,13 @@ void NVPTXPassConfig::addIRPasses() {
     addPass(createSROAPass());
   }
 
-  const auto &Options = getNVPTXTargetMachine().Options;
-  addPass(createNVPTXLowerUnreachablePass(Options.TrapUnreachable,
-                                          Options.NoTrapAfterNoreturn));
+  if (ST.hasPTXASUnreachableBug()) {
```
[the remark this answers is not in the document]
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXTargetMachine.cpp:371]: For PTX > 8.3 it's not needed.

For PTX versions older than that, we should assume that it *may* be given to the ptxas with a bug. The problem is that we make the decision on how to generate code at the build time, but we may not know which ptxas will be used until the run time, in general case. We should conservatively assume that we do need to produce `trap; exit;` in that case.

If we eventually find a legitimate use case for such an option, we can add it then. For now, I do not see much point.

File: llvm/test/CodeGen/NVPTX/unreachable.ll [hunk 2]
```diff
@@ -24,25 +26,30 @@ define void @kernel_func() {
   call void @throw()
 ; CHECK-TRAP-NOT: exit;
 ; CHECK-TRAP: trap;
+
 ; CHECK-NOTRAP-NOT: trap;
 ; CHECK: exit;
+
+; CHECK-BUG-FIXED-NOT: exit;
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/unreachable.ll:33]: Let's convert the test to use autogenerated checks.
The negative checks give me headaches and are often broken.

File: llvm/test/CodeGen/NVPTX/unreachable.ll [hunk 3]
```diff
@@ -24,25 +26,30 @@ define void @kernel_func() {
...
+; CHECK-BUG-FIXED: trap;
```
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/unreachable.ll:34]: E.g. what if we have `exit` *after* the `trap`?

**Review by Artem-B — APPROVED** [review]
LGTM with a test nit.

File: llvm/test/CodeGen/NVPTX/unreachable.ll (hunk 3 above)
Comment (justinfargnoli) [reply llvm/test/CodeGen/NVPTX/unreachable.ll]: 328e12c94e186dc8474646d96355ce4e928e8943

File: llvm/test/CodeGen/NVPTX/unreachable.ll (hunk 2 above)
Comment (justinfargnoli) [reply llvm/test/CodeGen/NVPTX/unreachable.ll]: 328e12c94e186dc8474646d96355ce4e928e8943

File: llvm/lib/Target/NVPTX/NVPTXSubtarget.h [hunk 4]
```diff
@@ -95,6 +95,7 @@ class NVPTXSubtarget : public NVPTXGenSubtargetInfo {
   bool hasDotInstructions() const {
     return SmVersion >= 61 && PTXVersion >= 50;
   }
+  bool hasPTXASUnreachableBug() const { return PTXVersion < 83; }
```
[the remark this answers is not in the document]
Comment (justinfargnoli) [reply llvm/lib/Target/NVPTX/NVPTXSubtarget.h:105]: 5d2902dde6a611b6dc531a0c747c596415326e30

**Review by Artem-B — APPROVED** [review]

**Review by jholewinski — APPROVED** [review]
Latest revision LGTM.

**Pull request merged**